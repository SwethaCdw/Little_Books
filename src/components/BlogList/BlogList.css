.blog-container {
  width: 85%;
  display: flex;
}
.blog-container .blog-list {
  width: 40%;
  background-color: #f5f6fa;
  padding: 1.875rem;
}
.blog-container .blog-list .blog-list-header {
  display: flex;
  align-items: normal;
  justify-content: space-between;
  gap: 0.625rem;
  margin-bottom: 1.25rem;
}
.blog-container .blog-list .blog-list-header .search-input {
  width: 75%;
  height: 50px;
  padding: 0 1.25rem;
  border-radius: 1.75rem;
  border: none;
  background-color: #ebecf3;
  font-size: 0.9375rem;
}
.blog-container .blog-list .blog-list-header .new-blog {
  width: 25%;
  border: none;
  border-radius: 2rem;
  color: white;
  font-weight: bold;
  font-size: 1.125rem;
}
.blog-container .blog-list .card-list {
  overflow: auto;
  height: 100%;
  padding-bottom: 3.75rem;
}
.blog-container .blog-list .card-list::-webkit-scrollbar {
  display: none;
}
.blog-container .blog-list .card-list .card {
  border-radius: 0.625rem;
  background-color: white;
}
.blog-container .blog-list .card-list .card.active {
  background-color: #96e2f4;
}
.blog-container .detailed-description {
  width: 57%;
  margin: 1.6rem auto;
}
.blog-container .detailed-description img {
  width: 94%;
  height: 450px;
  border-radius: 1.25rem;
  margin: 1.25rem 1.25rem 0;
  -o-object-fit: cover;
     object-fit: cover;
}
.blog-container .detailed-description .selected-blog-title {
  width: 100%;
  font-size: 1.75rem;
  font-family: "Roboto Slab", serif;
  padding: 0 0.625rem;
  margin: 1.5rem 0.625rem;
  border: none;
  font-size: 2.25rem;
  font-weight: 500;
}
.blog-container .detailed-description .selected-blog-details {
  height: 30%;
  padding: 0 0.625rem 1.25rem;
  margin: 0 0.625rem;
  overflow: auto;
  width: 100%;
  font-size: 1.1rem;
  line-height: 1.5;
  margin-bottom: 0.9375rem;
  border: none;
}
.blog-container .detailed-description .save-content-btn {
  background-color: #b354b7;
  color: white;
  margin-left: 0.625rem;
}
.blog-container .detailed-description button {
  margin-bottom: 1.875rem;
  float: right;
  border: none;
  padding: 0.625rem 1.5rem;
  border-radius: 1.875rem;
  background-color: #69f8f9;
  font-weight: 600;
}
.blog-container.dark {
  background-color: #021929;
  color: white;
}
.blog-container.dark .blog-list {
  background-color: black;
  color: white;
}
.blog-container.dark .blog-list .search-input {
  background-color: #021929;
}
.blog-container.dark .blog-list .card {
  background-color: #042b46;
  color: white;
}
.blog-container.dark .blog-list .card.active {
  background-color: #021929;
}
.blog-container.dark input, .blog-container.dark textarea {
  color: white;
  background-color: #021929;
}
.blog-container textarea {
  font-family: inherit;
}/*# sourceMappingURL=BlogList.css.map */