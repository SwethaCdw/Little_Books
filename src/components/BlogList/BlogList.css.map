{"version":3,"sources":["BlogList.scss","BlogList.css","../../styles/colors.scss","../../styles/mixins.scss"],"names":[],"mappings":"AAIA;EACI,UAAA;EACA,aAAA;ACHJ;ADII;EACI,UAAA;EACA,yBETU;EFUV,iBAAA;ACFR;ADGQ;EGVJ,aAAA;EACA,mBHU2C;EGT3C,8BHS4B;EACpB,aAAA;EACA,sBAAA;ACCZ;ADAY;EACI,UAAA;EACA,YAAA;EACA,kBAAA;EACA,sBAAA;EACA,YAAA;EACA,yBEpBK;EFqBL,oBAAA;ACEhB;ADAY;EACI,UAAA;EACA,YAAA;EACA,mBAAA;EACA,YE1BC;EF2BD,iBAAA;EACA,mBAAA;ACEhB;ADCQ;EACI,cAAA;EACA,YAAA;EACA,uBAAA;ACCZ;ADAY;EACI,aAAA;ACEhB;ADCY;EACI,uBAAA;EACA,uBExCH;ADyCb;ADCgB;EACI,yBE1CA;AD2CpB;ADII;EACI,UAAA;EACA,mBAAA;ACFR;ADGQ;EACI,UAAA;EACA,aAAA;EACA,sBAAA;EACA,yBAAA;EACA,oBAAA;KAAA,iBAAA;ACDZ;ADIQ;EACI,WAAA;EACA,kBAAA;EACA,iCAAA;EACA,mBAAA;EACA,uBAAA;EACA,YAAA;EACA,kBAAA;EACA,gBAAA;ACFZ;ADIQ;EACI,WAAA;EACA,2BAAA;EACA,kBAAA;EACA,cAAA;EACA,WAAA;EACA,iBAAA;EACA,gBAAA;EACA,wBAAA;EACA,YAAA;ACFZ;ADIQ;EACI,yBE9EI;EF+EJ,YEzEM;EF0EN,qBAAA;ACFZ;ADIQ;EACI,uBAAA;EACA,YAAA;EACA,YAAA;EACA,wBAAA;EACA,uBAAA;EACA,yBEtFG;EFuFH,gBAAA;ACFZ;ADKI;EACI,yBEzFU;EF0FV,YExFU;ADqFlB;ADIQ;EACI,uBEzFgB;EF0FhB,YE3FM;ADyFlB;ADGY;EACI,yBE/FE;AD8FlB;ADGY;EACI,yBEjGJ;EFkGI,YEjGE;ADgGlB;ADEgB;EACI,yBErGF;ADqGlB;ADIQ;EACI,YExGM;EFyGN,yBE3GM;ADyGlB;ADKI;EACI,oBAAA;ACHR","file":"BlogList.css"}