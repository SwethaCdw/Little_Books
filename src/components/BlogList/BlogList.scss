.blog-container {
    width: 85%;
    display: flex;
    .blog-list {
        width: 40%;
        background-color: #f5f6fa;
        padding: 30px;
        .blog-list-header {
            display: flex;
            justify-content: space-between;
            gap: 15px;
            margin-bottom: 20px;
            .search-input {
                width: 75%;
                height: 50px;
                padding: 0 20px;
                border-radius: 28px;
                border: none;
                background-color: #ebecf3;
            }
            .new-blog {
                width: 25%;
                border: none;
                border-radius: 32px;
                color: white;
                font-weight: bold;
                font-size: 17px;
            }
        }
        .card-list {
            overflow: auto;
            height: 100%;
            padding-bottom: 60px;
            &::-webkit-scrollbar {
                display: none;
              }

            .card{
                border-radius: 10px;
                background-color: #ffffff;

                &.selected {
                    background-color: #96e2f4;      
                }
            }
        }
    }
    .detailed-description {
        width: 57%;
        overflow: auto;
        margin: 25px auto;
        padding: 30px;
        img {
            width: 100%;
            height: 450px;
            border-radius: 20px;
            object-fit: cover;

        
        }
        .selected-blog-title {
            font-family: "Roboto Slab", serif;
            margin: 25px 0;
            border: none;
            font-size: 35px;
            font-weight: 800;
        }
        .selected-blog-details {
            font-family: "IBM Plex Sans", sans-serif;
            width: 100%;
            font-size: 17px;
            line-height: 1.5;
            margin-bottom: 15px;
            border: none;
        }
        .save-content-btn {
            background-color: #b354b7;
            color: white;
            margin-left: 10px;
        }
        button {
            margin-bottom: 30px;
            float: right;
            border: none;
            padding: 10px 25px;
            border-radius: 30px;
            background-color: #69f8f9;
        }
    }
    &.dark {
        background-color: #021929;
        color: white;
        .blog-list {
            background-color: black;
            color: white;
            .search-input {
                background-color: #021929 ;
            }
            .card {
                background-color: #042b46;
                color:white;
                &.selected {
                    background-color: #021929 ;      
                }
            }
        }
        input {
            color: white;
        }
       
    }
   
}