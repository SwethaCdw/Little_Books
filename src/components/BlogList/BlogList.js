import React, { useEffect, useRef, useState } from 'react';
import { fetchData } from '../../services/Api-service';
import Card from '../Card/Card';
import './BlogList.css';
import Search from '../Search/Search';
import { handleImageError } from '../../utils/common-utils';
import { FALLBACK_IMAGE } from '../../constants/common-constants';
import { useDispatch, useSelector } from 'react-redux';
import { setBlogsData, setSearchTerm, setSelectedBlog, setImageLoaded, setShowNewBlogModal, updateSelectedBlog } from '../../store/blogSlice';
import NewBlogModal from '../NewBlog/NewBlog';
import AlertDialog from '../Dialog/AlertDialog';
import { getItemFromLocalStorage } from '../../utils/local-storage-utils';


const BlogList = () => {
    const dispatch = useDispatch();
    const blogsData = useSelector((state) => state.blog.blogsData);
    const searchTerm = useSelector((state) => state.blog.searchTerm);
    const selectedBlog = useSelector((state) => state.blog.selectedBlog);
    const imageLoaded = useSelector((state) => state.blog.imageLoaded);
    const showNewBlogModal = useSelector((state) => state.blog.showNewBlogModal);
    const darkMode = useSelector((state) => state.theme.darkMode);
    const selectedFilters = useSelector((state) => state.blog.selectedFilters);
    const [editMode, setEditMode] = useState(false);
    const [showWarningModal, setShowWarningModal] = useState(false);
    const [clickedBlog, setClickedBlog] = useState(null);
    const [textareaHeight, setTextareaHeight] = useState('auto');
    const textareaRef = useRef(null); 

    const handleNewBlogClick = () => {
      dispatch(setShowNewBlogModal(true));
    };
  
    const handleCloseModal = () => {
      dispatch(setShowNewBlogModal(false));
    };

    const adjustTextareaHeight = () => {
      if (textareaRef?.current) {
        setTextareaHeight(`${textareaRef?.current.scrollHeight}px`);
      }
    };

    const handleEditContent = () => {
      setEditMode(!editMode);
      if (editMode) {
        adjustTextareaHeight(); 
      }
    };  

    const handleSaveContent = () => {
      dispatch(updateSelectedBlog(selectedBlog));
      setEditMode(false); 
    };

    useEffect(() => {
      const fetchDataFromApi = async () => {
        try {
          const storedBlogs = JSON.parse(getItemFromLocalStorage('blogs'));
          if(storedBlogs?.length){
            dispatch(setBlogsData(storedBlogs));
            dispatch(setSelectedBlog(storedBlogs[0]));
          } else {
            const data = await fetchData('resources/blogs.json');
            dispatch(setBlogsData(data));
            dispatch(setSelectedBlog(data[0]));
          }
        } catch (error) {
          console.error('Error fetching data:', error);
        }
      };

      fetchDataFromApi();
    }, [dispatch]);

    const handleImageLoaded = () => {
      dispatch(setImageLoaded(true));
    };

    const filteredBlogs = blogsData?.filter(
      (item) => 
        item.title.toLowerCase().includes(searchTerm.toLowerCase()) ||
        item.details.toLowerCase().includes(searchTerm.toLowerCase()) ||
        item.type.toLowerCase().includes(searchTerm.toLowerCase())
    ).filter((item) => selectedFilters.includes(item.type));;

    const handleBlogClick = (blogDetails) => {
      if(editMode){
        setClickedBlog(blogDetails);

        setShowWarningModal(true);
      } else {
        dispatch(setSelectedBlog(blogDetails));
      }
    }

    const handleAlertClose = () => {
      setShowWarningModal(false);
    }

    const handleOkay = () => {
      setShowWarningModal(false);
      setEditMode(!editMode);
      dispatch(setSelectedBlog(clickedBlog));

    }


  return (
    <div className={`blog-container ${darkMode ? 'dark' : ''}`}>
        <div className='blog-list'>
            <div className='blog-list-header'>
                <Search 
                  className='search-input' 
                  value={searchTerm} 
                  onChange={(e) => dispatch(setSearchTerm(e.target.value))} 
                  placeholder='Search Blog' 
                  type='text'
                />
                <button className='new-blog' onClick={handleNewBlogClick}>NEW</button>
            </div>
            <ul className='card-list'>
                {filteredBlogs.length ? ( 
                  filteredBlogs?.map(item => {
                    return (
                        <li key={item.title} className={`card ${selectedBlog && selectedBlog.title === item.title ? 'selected' : ''}`} onClick={() => handleBlogClick(item)}>
                            <Card 
                                key={item.title}
                                title={item.title}
                                type={item.type}
                                description={item.details}
                            />
                         </li>
                    )
                })) : (
                  <p className='no-blogs-found'>No blogs found.</p>
                )}
            </ul>
        </div>
        <div className='detailed-description'>
            {selectedBlog &&
            <>
                {!imageLoaded && <div className="loader">Loading...</div>}
                <img
                    src={selectedBlog.photo || FALLBACK_IMAGE}
                    alt={selectedBlog.title}
                    style={{ display: imageLoaded ? 'block' : 'none' }}
                    onLoad={handleImageLoaded}
                    onError={handleImageError}
                />
                {editMode ? ( 
                  <>
                    <input type='text' className='selected-blog-title' value={selectedBlog.title} onChange={(e) => dispatch(setSelectedBlog({ ...selectedBlog, title: e.target.value }))}></input>
                    <textarea
                      ref={textareaRef} 
                      value={selectedBlog.details}
                      onChange={(e) => dispatch(setSelectedBlog({ ...selectedBlog, details: e.target.value }))}
                      className='selected-blog-details'
                      style={{ height: textareaHeight }}
                      />
                    <button onClick={handleSaveContent} className='save-content-btn'>SAVE CONTENT</button>
                  </>
                ) : (
                  <>
                    <h1 className='selected-blog-title'>{selectedBlog.title}</h1>
                    <p className='selected-blog-details'>{selectedBlog.details}</p>
                  </>
                )}
                
                <button onClick={handleEditContent}>
                  {editMode ? 'CANCEL' : 'EDIT CONTENT'}
                </button>
            </>  
            }
        </div>
        {showNewBlogModal && <NewBlogModal onClose={handleCloseModal} />}
        {showWarningModal && <AlertDialog title='Do you wish stop the edit?' cancelText='Cancel' okayText='Yes' handleAlertClose={handleAlertClose} handleOkay={handleOkay}/>}

    </div>
  );
};

export default BlogList;
