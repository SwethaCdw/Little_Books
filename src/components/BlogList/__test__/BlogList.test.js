import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import BlogList from '../BlogList';

const mockStore = configureStore([]);

describe('BlogList component', () => {
  let store;
  beforeEach(() => {
    store = mockStore({
      blog: {
        blogsData: [
          { title: 'Blog 1', type: 'Type 1', details: 'Details 1' },
          { title: 'Blog 2', type: 'Type 2', details: 'Details 2' },
        ],
        searchTerm: '',
        selectedBlog: { title: 'Blog 1', type: 'Type 1', details: 'Details 1' },
        imageLoaded: true,
        showNewBlogModal: false,
        selectedFilters: ['Type 1', 'Type 2'],
      },
      theme: {
        darkMode: false,
      },
    });
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(<Provider store={store}><BlogList /></Provider>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the blog list correctly', () => {
    render(
      <Provider store={store}>
        <BlogList />
      </Provider>
    );

    expect(screen.getByText('Blog 2')).toBeInTheDocument();
  });

  it('should handle blog click', () => {
    render(
      <Provider store={store}>
        <BlogList />
      </Provider>
    );

    const blogItem = screen.getByText('Blog 2');

    fireEvent.click(blogItem);

    const expectedAction = { type: 'blog/setSelectedBlog', payload: { title: 'Blog 2', type: 'Type 2', details: 'Details 2' } };
    expect(store.getActions()).toContainEqual(expectedAction);
  });

  it('displays no blogs found message when filtered blogs are empty', () => {
    const emptyStore = mockStore({
      blog: {
        blogsData: [],
        searchTerm: '',
        selectedBlog: null,
        imageLoaded: true,
        showNewBlogModal: false,
        selectedFilters: [],
      },
      theme: {
        darkMode: false,
      },
    });
    render(
      <Provider store={emptyStore}>
        <BlogList />
      </Provider>
    );
    expect(screen.getByText('No blogs found.')).toBeInTheDocument();
  });



 
});
