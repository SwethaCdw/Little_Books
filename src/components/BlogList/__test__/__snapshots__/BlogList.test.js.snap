// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogList component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="blog-container "
  >
    <div
      class="blog-list"
    >
      <div
        class="blog-list-header"
      >
        <input
          class="search-input"
          placeholder="Search Blog"
          type="text"
          value=""
        />
        <button
          class="new-blog"
        >
          NEW
        </button>
      </div>
      <ul
        class="card-list"
      >
        <li
          class="card selected"
        >
          <div
            class="card-container"
          >
            <h3
              class="title"
            >
              Blog 1
            </h3>
            <p
              class="type"
            >
              TYPE 1
            </p>
            <p
              class="description"
              title="DETAILS 1"
            >
              Details 1
            </p>
          </div>
        </li>
        <li
          class="card "
        >
          <div
            class="card-container"
          >
            <h3
              class="title"
            >
              Blog 2
            </h3>
            <p
              class="type"
            >
              TYPE 2
            </p>
            <p
              class="description"
              title="DETAILS 2"
            >
              Details 2
            </p>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="detailed-description"
    >
      <img
        alt="Blog 1"
        src="https://clarionhealthcare.com/wp-content/uploads/2020/12/default-fallback-image.png"
        style="display: block;"
      />
      <h1
        class="selected-blog-title"
      >
        Blog 1
      </h1>
      <p
        class="selected-blog-details"
      >
        Details 1
      </p>
      <button>
        EDIT CONTENT
      </button>
    </div>
  </div>
</DocumentFragment>
`;
